test/test_kvtree_write_gather.c

```diff
@@ -0,0 +1,59 @@
+#include "test_kvtree.h"
+#include "test_kvtree_util.h"
+#include "mpi.h"
+#include "stdio.h"
+
+
+int main(int argc, char** argv){
+  kvtree* recv;
+  int rc = TEST_PASS;
+  int rank, ranks, kvtree_rc;
+  int val_of_one, val_of_two, val_of_three;
+  const char one[3] = "ONE", two[3] = "TWO", three[5] = "THREE";
+  MPI_Init(&argc, &argv);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &ranks);
+
+  kvtree* kvtree = kvtree_new();
+  recv = kvtree_new();
+  if(rank == 0){
+    kvtree_util_set_int(kvtree, "ONE", 1);
+    kvtree_util_set_int(kvtree, "TWO", 2);
+  }
+  if(rank == 2){
+    kvtree_util_set_int(kvtree, "ONE", -1);
+    kvtree_util_set_int(kvtree, "THREE", 3);
+  }
+
+  kvtree_rc = kvtree_write_gather("PREV_", kvtree, MPI_COMM_WORLD);
+  if (kvtree_rc != KVTREE_SUCCESS) rc = TEST_FAIL;
+  if (kvtree == NULL) rc = TEST_FAIL;
+  kvtree_rc = kvtree_read_scatter("PREV_", recv, MPI_COMM_WORLD);
+  if (kvtree_rc != KVTREE_SUCCESS) rc = TEST_FAIL;
+  if (kvtree == NULL) rc = TEST_FAIL;
+  if(rank == 0){
+    kvtree_print(kvtree, 4);
+    kvtree_print(recv, 4);
+    kvtree_rc = kvtree_util_get_int(kvtree, "ONE", &val_of_one);
```

Comment:
Do you mean to be checking "recv" instead of "kvtree" after the read_scatter?

Reply:
All of the checks for which keys are set after the read_scatter are against the tree used in the write_gather step.  I was wondering whether you really meant to check against the tree used in the read_scatter instead.